src/hotspot/share/gc/g1/g1BlockOffsetTable.cpp

```diff
@@ -211,55 +211,34 @@ void G1BlockOffsetTable::update_for_block_work(HeapWord* blk_start, HeapWord* bl
-  }
 #endif
+
+  debug_only(G1BlockOffsetTable::verify_for_block(blk_start, blk_end);)
 }
 
-void G1BlockOffsetTable::verify(const HeapRegion* hr) const {
-  assert(hr->bottom() < hr->top(), "Only non-empty regions should be verified.");
-  uint8_t* start_card = entry_for_addr(hr->bottom());
-  uint8_t* end_card = entry_for_addr(hr->top() - 1);
+void G1BlockOffsetTable::verify_for_block(HeapWord* blk_start, HeapWord* blk_end) const {
+  assert(is_crossing_card_boundary(blk_start, blk_end), "precondition");
 
-  for (uint8_t* current_card = start_card; current_card < end_card; current_card++) {
-    uint8_t entry = offset_array(current_card);
-    if (entry < CardTable::card_size_in_words()) {
-      // The entry should point to an object before the current card. Verify that
-      // it is possible to walk from that object in to the current card by just
-      // iterating over the objects following it.
-      HeapWord* card_address = addr_for_entry(current_card);
-      HeapWord* obj_end = card_address - entry;
-      while (obj_end < card_address) {
-        HeapWord* obj = obj_end;
-        size_t obj_size = hr->block_size(obj);
-        obj_end = obj + obj_size;
-        guarantee(obj_end > obj && obj_end <= hr->top(),
-                  "Invalid object end. obj: " PTR_FORMAT " obj_size: " SIZE_FORMAT " obj_end: " PTR_FORMAT " top: " PTR_FORMAT,
-                  p2i(obj), obj_size, p2i(obj_end), p2i(hr->top()));
-      }
-    } else {
-      // Because we refine the BOT based on which cards are dirty there is not much we can verify here.
-      // We need to make sure that we are going backwards and that we don't pass the start of the
-      // corresponding heap region. But that is about all we can verify.
-      size_t backskip = BOTConstants::entry_to_cards_back(entry);
-      guarantee(backskip >= 1, "Must be going back at least one card.");
+  uint8_t* start_card = entry_for_addr(align_up_by_card_size(blk_start));
+  uint8_t* end_card = entry_for_addr(blk_end - 1);
+  // Check cards in [start_card, end_card]
+  assert(offset_array(start_card) < CardTable::card_size_in_words(), "offset card");
 
-      size_t max_backskip = current_card - start_card;
-      guarantee(backskip <= max_backskip,
-                "Going backwards beyond the start_card. start_card: " SIZE_FORMAT " current_card: " SIZE_FORMAT " backskip: " SIZE_FORMAT,
-                p2i(start_card), p2i(current_card), backskip);
+  for (uint8_t* current_card = start_card + 1; current_card <= end_card; ++current_card) {
+    assert(offset_array(current_card) > 0 &&
+           offset_array(current_card) <= (uint8_t) (CardTable::card_size_in_words() + BOTConstants::N_powers - 1),
+           "offset array should have been set - "
+           "%u not > 0 OR %u not <= %u",
+           (uint) offset_array(current_card),
+           (uint) offset_array(current_card),
+           (uint) (CardTable::card_size_in_words() + BOTConstants::N_powers - 1));
```

Comment:
The first assert could have the same descriptive assert message than the other; define a static method that gets the upper bound passed.

Reply:
I moved the common content to a new method. But I don't mark it as `static` because the `offset_array` it uses is not `static`.